presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java

```diff
@@ -0,0 +1,282 @@
+        }
+        else if (columnType == SMALLINT) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE, FieldType.SHORT);
+        }
+        else if (columnType == TINYINT) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE);
+        }
+        else if (columnType == BOOLEAN) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE, FieldType.SHORT, FieldType.INT, FieldType.LONG);
+        }
+        else if (columnType == DOUBLE) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.DOUBLE, FieldType.FLOAT);
+        }
+        else if (isVarcharType(columnType)) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE);
+        }
+    }
+
+    private void checkFieldTypeOneOf(FieldType declaredFieldType, String columnName, Type columnType, FieldType... allowedFieldTypes)
+    {
+        if (!Arrays.asList(allowedFieldTypes).contains(declaredFieldType)) {
+            throw new IllegalArgumentException(format(
+                    "Wrong dataFormat '%s' specified for column '%s'; %s type implies use of %s",
+                    declaredFieldType.name(),
+                    columnName,
+                    columnType.getDisplayName(),
+                    Joiner.on("/").join(allowedFieldTypes)));
+        }
+    }
+
+    private int parseMapping(EncoderColumnHandle columnHandle, FieldType fieldType)
+    {
+        String mapping = Optional.ofNullable(columnHandle.getMapping()).orElse("0");
+        Matcher mappingMatcher = MAPPING_PATTERN.matcher(mapping);
+        if (!mappingMatcher.matches()) {
+            throw new IllegalArgumentException(format("Invalid mapping format '%s' for column '%s'", mapping, columnHandle.getName()));
+        }
+        int start = parseInt(mappingMatcher.group(1));
+        OptionalInt end;
+        if (mappingMatcher.group(2) != null) {
+            end = OptionalInt.of(parseInt(mappingMatcher.group(2)));
+        }
+        else {
+            if (!isVarcharType(columnHandle.getType())) {
+                end = OptionalInt.of(start + fieldType.getSize());
+            }
+            else {
+                end = OptionalInt.empty();
+            }
+        }
+
+        checkArgument(start >= 0, "Start offset %s for column '%s' must be greater or equal 0", start, columnHandle.getName());
+        end.ifPresent(endValue -> {
+            checkArgument(endValue >= 0, "End offset %s for column '%s' must be greater or equal 0", endValue, columnHandle.getName());
+            checkArgument(endValue > start, "End offset %s for column '%s' must greater than start offset", endValue, columnHandle.getName());
+        });
+
+        int length = end.isPresent() ? end.getAsInt() - start : fieldType.getSize();
+
+        if (!isVarcharType(columnHandle.getType())) {
```

Comment:
it does not seem to handle case where we are dealing with varchar and `end` is not explicitly specified.
Would lenght be `1` then?

Reply:
The way I had the encoders structured before didn't let me 'see' the value of the mapping for the next column so `varchar` types would need to specify the start and the end every time. Now I think I might be able to dynamically allocate space for varchar types by 'peaking' into the mapping for the next column and using the `start index - 1` for the next column as the end index for the varchar column. I could also use `VarcharType#getLength` if there isn't a next column.

Reply:
that's how the decoder works rn afaik

Reply:
So that means that a missing end value is legal only when the field is in the last position.

Reply:
Yes, I will have to figure out a way to allocate more space in the `ByteBuffer` if `end` is not specified for `VARCHAR`.